File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/usecase/external/erp/usecases/GetProduct.java
package com.mycompany.myshop.testkit.dsl.core.usecase.external.erp.usecases;

import com.mycompany.myshop.testkit.driver.port.external.erp.ErpDriver;
import com.mycompany.myshop.testkit.driver.port.external.erp.dtos.GetProductRequest;
import com.mycompany.myshop.testkit.driver.port.external.erp.dtos.GetProductResponse;
import com.mycompany.myshop.testkit.dsl.core.usecase.external.erp.usecases.base.BaseErpUseCase;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseResult;
import com.mycompany.myshop.testkit.dsl.core.shared.UseCaseContext;
import com.mycompany.myshop.testkit.driver.port.dtos.error.SystemError;

public class GetProduct extends BaseErpUseCase<GetProductResponse, GetProductVerification> {
    private String skuParamAlias;

    public GetProduct(ErpDriver driver, UseCaseContext context) {
        super(driver, context);
    }

    public GetProduct sku(String skuParamAlias) {
        this.skuParamAlias = skuParamAlias;
        return this;
    }

    @Override
    public UseCaseResult<GetProductResponse, GetProductVerification> execute() {
        var sku = context.getParamValue(skuParamAlias);

        var request = GetProductRequest.builder()
                .sku(sku)
                .build();

        var result = driver.getProduct(request);

        return new UseCaseResult<>(result.mapError(e -> SystemError.of(e.getMessage())), context, GetProductVerification::new);
    }
}

File: system-test/typescript/src/testkit/dsl/core/usecase/external/erp/usecases/GetProduct.ts
import type { ErpDriver } from '../../../../../../driver/port/external/erp/erp-driver.js';
import type { GetProductRequest } from '../../../../../../driver/port/external/erp/dtos/GetProductRequest.js';
import type { GetProductResponse } from '../../../../../../driver/port/external/erp/dtos/GetProductResponse.js';
import { success, failure, type Result } from '../../../../../../common/result.js';
import type { SystemError } from '../../../../../../driver/port/dtos/errors/SystemError.js';
import { UseCaseResult } from '../../../../shared/use-case-result.js';
import type { UseCaseContext } from '../../../../shared/use-case-context.js';
import { BaseErpUseCase } from './base/BaseErpUseCase.js';
import { GetProductVerification } from './GetProductVerification.js';

export class GetProduct extends BaseErpUseCase<GetProductResponse, GetProductVerification> {
  private _skuParamAlias?: string;

  constructor(driver: ErpDriver, context: UseCaseContext) {
    super(driver, context);
  }

  sku(skuParamAlias: string): this {
    this._skuParamAlias = skuParamAlias;
    return this;
  }

  async execute(): Promise<UseCaseResult<GetProductResponse, GetProductVerification>> {
    const sku = this.context.getParamValue(this._skuParamAlias);

    const request: GetProductRequest = {
      sku: sku ?? '',
    };

    const driverResult = await this.driver.getProduct(request);
    const result: Result<GetProductResponse, SystemError> = driverResult.success
      ? success(driverResult.value)
      : failure({ message: driverResult.error.message ?? 'ERP error', fieldErrors: [] });

    return new UseCaseResult(
      result,
      this.context,
      (response, ctx) => new GetProductVerification(response, ctx),
    );
  }
}
